numpy/_core/src/npysort/binsearch.cpp

```diff
@@ -64,43 +64,100 @@ binsearch(const char *arr, const char *key, char *ret, npy_intp arr_len,
+    against the same set of pivots. For example, in the first iteration all
+    keys are compared against the median. In the second iteration, all keys
+    end up being compared against 1st and 3rd quartiles.
+
+    2. Independent calculations for out-of-order execution. In the single-key
+    version, step i+1 depends on computation of step i. Meaning that step i+1
+    must wait for step i to complete before proceeding. When batching multiple
+    keys, we compute each step for all keys before continuing on the next
+    step. All the computations at a given step are independent across
+    different keys. Meaning that the CPU can execute multiple keys
+    out-of-order in parallel.
+
+    Invariant:
+    - cmp(arr[i], key_val) == true  for all i < base
+    - cmp(arr[i], key_val) == false for all i >= base + length
+
+    The insertion index candidates are i in range [base, base+length] and
+    on each iteration we shrink the range into either
+        [base, ceil(length / 2)]
+    or
+        [base + floor(length / 2), ceil(length / 2)]
+
+    Optimization: we unroll the first iteration for the following reasons:
+        1. ret is not initialized with the bases, so we save |keys| writes
+        by not having to intialize it with 0s.
+        2. By assuming the initial base for every key is 0, we also save
+        |keys| reads.
+        3. In the first iteration, all elements are compared against the
+        median. So we can store it in a variable and use it for all keys.
+
+    This initial block replaces the initialization loop that is used for the
+    arr_len==0 case. Note that when arr_len = 1, then half is 0 so the
+    following block initializes the array as with 0s.
+    */
+    npy_intp interval_length = arr_len;
+    npy_intp half = interval_length >> 1;
+    interval_length -= half; // length -> ceil(length / 2)
+
+    npy_intp base = 0;
+    const T mid_val = *(const T *)(arr + (base + half) * arr_str);
+
+    for (npy_intp i = 0; i < key_len; ++i) {
+        const T key_val = *(const T *)(key + i * key_str);
+        *(npy_intp *)(ret + i * ret_str) = cmp(mid_val, key_val) * half;
+    }
 
-        last_key_val = key_val;
+    while (interval_length > 1) {
+        npy_intp half = interval_length >> 1;
+        interval_length -= half; // length -> ceil(length / 2)
 
-        while (min_idx < max_idx) {
-            const npy_intp mid_idx = min_idx + ((max_idx - min_idx) >> 1);
-            const T mid_val = *(const T *)(arr + mid_idx * arr_str);
-            if (cmp(mid_val, key_val)) {
-                min_idx = mid_idx + 1;
-            }
-            else {
-                max_idx = mid_idx;
-            }
+        for (npy_intp i = 0; i < key_len; ++i) {
+            npy_intp &base = *(npy_intp *)(ret + i * ret_str);
+            const T mid_val = *(const T *)(arr + (base + half) * arr_str);
+            const T key_val = *(const T *)(key + i * key_str);
+            base += cmp(mid_val, key_val) * half;
```

Comment:
This requires a multiplication. Maybe 
```suggestion
            if cmp(mid_val, key_val):
               base += half;
```
is faster.

Reply:
I structured it in that way to avoid branches because they are very hard to predict (they are 50/50 on random input). However when I checked what the compiler was doing it actually translated it into a considional select and removed the multiplication.